bootstrap/controllers/certificates_controller.go

```diff
@@ -0,0 +1,219 @@
+	if err != nil {
+		log.Error(err, "Failed to get config owner")
+		return ctrl.Result{}, err
+	}
+	if configOwner == nil {
+		return ctrl.Result{}, nil
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, m.GetNamespace(), m.Spec.ClusterName)
+
+	microclusterPort := config.Spec.ControlPlaneConfig.GetMicroclusterPort()
+	workload, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(cluster), microclusterPort)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	patchHelper, err := patch.NewHelper(m, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create patch helper for machine: %w", err)
+	}
+
+	scope := &CertificatesScope{
+		Log:      log,
+		Machine:  m,
+		Config:   config,
+		Cluster:  cluster,
+		Patcher:  patchHelper,
+		Workload: workload,
+	}
+
+	if refreshCertificates {
+		if configOwner.IsControlPlaneMachine() {
+			return r.refreshControlPlaneCertificates(ctx, scope)
+		} else {
+			return ctrl.Result{}, fmt.Errorf("Worker nodes are not supported yet")
+		}
+	}
+
+	if !updateExpiryDate {
+		return r.updateExpiryDateAnnotation(ctx, scope)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *CertificatesReconciler) refreshControlPlaneCertificates(ctx context.Context, scope *CertificatesScope) (reconcile.Result, error) {
+	nodeToken, err := token.LookupNodeToken(ctx, r.Client, util.ObjectKey(scope.Cluster), scope.Machine.Name)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	mAnnotations := scope.Machine.GetAnnotations()
+
+	refreshAnnotation, ok := mAnnotations[bootstrapv1.CertificatesRefreshAnnotation]
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	seconds, err := utiltime.TTLToSeconds(refreshAnnotation)
+	scope.Log.Info("Refreshing Certificates", "TTL", refreshAnnotation, "seconds", seconds)
+	if err != nil {
+		scope.Log.Error(err, "Failed to parse TTL annotation")
+		return ctrl.Result{}, err
+	}
+
+	controlPlaneConfig := scope.Config.Spec.ControlPlaneConfig
+	controlPlaneEndpoint := scope.Cluster.Spec.ControlPlaneEndpoint.Host
+	extraSANs := append(controlPlaneConfig.ExtraSANs, controlPlaneEndpoint)
+
+	expirySecondsUnix, err := scope.Workload.RefreshCertificates(ctx, scope.Machine, *nodeToken, seconds, extraSANs)
+	if err != nil {
+		scope.Log.Error(err, "Failed to refresh certificates")
+		r.recorder.Eventf(scope.Machine, corev1.EventTypeWarning, bootstrapv1.CertificatesRefreshFailedEvent, "Failed to refresh certificates: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	scope.Log.Info("Certificates refreshed", "expiry", expirySecondsUnix)
+
+	expiryTime := time.Unix(int64(expirySecondsUnix), 0)
+
+	delete(mAnnotations, bootstrapv1.CertificatesRefreshAnnotation)
+	mAnnotations[bootstrapv1.MachineCertificatesExpiryDateAnnotation] = expiryTime.Format(time.RFC3339)
+	scope.Machine.SetAnnotations(mAnnotations)
+	if err := scope.Patcher.Patch(ctx, scope.Machine); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch machine annotations: %w", err)
+	}
```

Comment:
Just an idea... how about requeuing if this failed because of a concurrent update?

https://sourcegraph.com/search?q=context:global+apierrors.IsConflict&patternType=keyword&sm=0

Reply:
In this case, if the returned error is non-nil, the Result is ignored and the request will be requeued using exponential backoff. The only way to stop this from reconciling is to return a TerminalError, so in this controller, everything will be retried using the exponential backoff mechanism.